lib/Internal/Yielder.php

```diff
@@ -120,25 +130,28 @@ private function yield($value): Promise
             throw new \Error("Streams cannot yield values after calling complete");
         }
 
-        if (\is_null($value)) {
-            throw new \TypeError("Streams cannot yield null");
-        }
-
         if ($value instanceof Promise || $value instanceof ReactPromise) {
             throw new \TypeError("Streams cannot yield promises");
         }
 
+        if ($key === null) {
+            $key = $this->nextKey++;
+        } elseif (\is_int($key) && $key >= $this->nextKey) {
+            $this->nextKey = $key + 1;
+        }
```

Comment:
IMO we should only allow all keys being `null` or none, as a mixture is very likely to cause bugs.

Reply:
Do you mean just auto-generate all keys?

Reply:
I mean either all yields must provide a non-null key or every yield must pass `null` per stream.

Reply:
Or do you rather mean once a custom key is used, error if a custom key is not provided. That definitely makes sense.

Reply:
Now that I think about it more, these should probably be two different data structures. It's like a list vs. map. If I filter a list, I expect the indices to be without gaps.

Reply:
I honestly don't think keys are very useful here, which is why I didn't include them in the first place. I think forbidding `null` made more sense.

Reply:
Updated to only auto-generate keys.